include/boost/url/impl/url_base.ipp

```diff
@@ -1337,6 +1337,166 @@ end() const noexcept ->
+segments_type::
+insert_encoded_segment_( segments_type::iterator pos, string_view s ) ->
+    iterator
+{
+    BOOST_ASSERT(detail::pchar_pct_set().check(s));
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+    BOOST_ASSERT(pos.v_ == &v);
+    BOOST_ASSERT(pos.off_ >= v.pt_.offset[detail::id_path]);
+    BOOST_ASSERT(pos.off_ <= v.pt_.offset[detail::id_query]);
+    auto const n0 = v.pt_.offset[detail::id_end];
+    auto const n = s.size() + 1;
+    v.s_ = v.a_.resize(v.size() + n);
+    v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) + n);
+    std::memmove(v.s_ + v.pt_.offset[detail::id_end] + pos.off_ - n0, v.s_ + pos.off_, n0 - pos.off_ + 1);
+    BOOST_ASSERT(v.s_[v.pt_.offset[detail::id_end]] == '\0');
+    v.s_[pos.off_] = '/';
+    std::memcpy(v.s_ + pos.off_ + 1, s.data(), s.size());
+    ++v.pt_.nseg;
+    pos.off_ += n;
+    pos.parse();
+    return pos;
+}
+
+auto
+url_base::
+segments_type::
+insert_segment_( segments_type::iterator pos, string_view s, std::size_t const ns ) ->
+    iterator
+{
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+    BOOST_ASSERT(pos.v_ == &v);
+    BOOST_ASSERT(pos.off_ >= v.pt_.offset[detail::id_path]);
+    BOOST_ASSERT(pos.off_ <= v.pt_.offset[detail::id_query]);
+    auto const pct = detail::pchar_pct_set();
+    BOOST_ASSERT(pct.encoded_size(s) == ns);
+    auto const n0 = v.pt_.offset[detail::id_end];
+    auto const n = ns + 1;
+    v.s_ = v.a_.resize(v.size() + n);
+    v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) + n);
+    std::memmove(v.s_ + v.pt_.offset[detail::id_end] + pos.off_ - n0, v.s_ + pos.off_, n0 - pos.off_ + 1);
+    BOOST_ASSERT(v.s_[v.pt_.offset[detail::id_end]] == '\0');
+    v.s_[pos.off_] = '/';
+    pct.encode(v.s_ + pos.off_ + 1, s);
+    ++v.pt_.nseg;
+    pos.off_ += n;
+    pos.parse();
+    return pos;
+}
+
+auto
+url_base::
+segments_type::
+insert_encoded_segment( segments_type::iterator pos, string_view s ) ->
+    iterator
+{
+    return insert_encoded_segment_(pos, detail::pchar_pct_set().validate(s));
+}
+
+auto
+url_base::
+segments_type::
+insert_segment( segments_type::iterator pos, string_view s ) ->
+    iterator
+{
+    return insert_segment_(pos, s, detail::pchar_pct_set().encoded_size(s));
+}
+
+auto
+url_base::
+segments_type::
+replace_encoded_segment( segments_type::iterator pos, string_view s ) ->
+    iterator
+{
+    detail::pchar_pct_set().validate(s);
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+    auto const ns = s.size();
+    auto const n0 = pos.n_;
+    auto const n = ns + 1;
+    if( n0 < n )
+        v.s_ = v.a_.reserve(v.a_.size() + n - n0);
+    auto const cap = v.a_.capacity();
+    auto r = insert_encoded_segment_(erase(pos), s);
+    BOOST_ASSERT(v.a_.capacity() == cap); // Strong guarantee violation
```

Comment:
`capacity` does not have to be preserved on exception

Reply:
You're not reading this correctly. We reserve first, this assert verifies that `insert_encoded_segment_` did not need to expand the memory, which would mean that it could fail, which would violate the strong guarantee.

Reply:
the coverage report will have no way to indicate that both paths of the `if` were covered